kotlinx-coroutines-core/common/src/flow/Builders.kt

```diff
@@ -315,3 +318,36 @@ private class ChannelFlowBuilder<T>(
     override fun toString(): String =
         "block[$block] -> ${super.toString()}"
 }
+
+private class CallbackFlowBuilder<T>(
+    private val block: suspend ProducerScope<T>.() -> Unit,
+    context: CoroutineContext = EmptyCoroutineContext,
+    capacity: Int = BUFFERED
+) : ChannelFlowBuilder<T>(block, context, capacity) {
+
+    private val collectCallback: suspend (ProducerScope<T>) -> Unit = {
+        collectTo(it)
+        /*
+         * We expect user either call `awaitClose` from within a block (then the channel is closed at this moment)
+         * or being closed/cancelled externally/manually. Otherwise "user forgot to call
+         * awaitClose and receives unhelpful ClosedSendChannelException exceptions" situation is detected.
+         */
+        if (it.isActive && !it.isClosedForSend) {
```

Comment:
Not sure about this very condition, will think it out for a while

Reply:
It seems that `!it.isClosedForSend` should be enough.